multiple rlib candidates passed to rustdoc

Starting with Rust 1.27, we're seeing duplicate rlibs passed to rustdoc for the `miniz_oxide_c_api` crate.
https://kojipkgs.fedoraproject.org/work/tasks/1896/27791896/build.log

Rust 1.26 was fine, and I can reproduce the issue up through nightly.

```
$ cargo +nightly -Vv
cargo 1.28.0-nightly (e2348c2db 2018-06-07)
release: 1.28.0
commit-hash: e2348c2db296ce33428933c3ab8786d5f3c54a2e
commit-date: 2018-06-07
```

```
$ rustc +nightly -Vv
rustc 1.28.0-nightly (01cc982e9 2018-06-24)
binary: rustc
commit-hash: 01cc982e936120acb0424e41de14e42ba2d88c6f
commit-date: 2018-06-24
host: x86_64-unknown-linux-gnu
release: 1.28.0-nightly
LLVM version: 6.0
```

It's easy to reproduce by checking out https://github.com/Frommi/miniz_oxide, for the current master at commit 10c070c61b2d412358196db1e83f48f3b7422d7e.

```
$ cargo +nightly build --release --verbose
   Compiling build_const v0.2.1
   Compiling cc v1.0.17
   Compiling libc v0.2.42
   Compiling adler32 v1.0.3
     Running `rustc --crate-name build_const /home/jistone/.cargo/registry/src/github.com-1ecc6299db9ec823/build_const-0.2.1/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 --cfg 'feature="default"' --cfg 'feature="std"' -C metadata=237bc424c8fc27b7 -C extra-filename=-237bc424c8fc27b7 --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --cap-lints allow`
     Running `rustc --crate-name libc /home/jistone/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.42/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C panic=abort --cfg 'feature="default"' --cfg 'feature="use_std"' -C metadata=583e369a8ae79b43 -C extra-filename=-583e369a8ae79b43 --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --cap-lints allow`
     Running `rustc --crate-name adler32 /home/jistone/.cargo/registry/src/github.com-1ecc6299db9ec823/adler32-1.0.3/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C panic=abort -C metadata=6582e487cd20c9f9 -C extra-filename=-6582e487cd20c9f9 --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --cap-lints allow`
     Running `rustc --crate-name cc /home/jistone/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.17/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C metadata=8386dfd55958a0e1 -C extra-filename=-8386dfd55958a0e1 --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --cap-lints allow`
   Compiling crc v1.8.1
     Running `rustc --crate-name build_script_build /home/jistone/.cargo/registry/src/github.com-1ecc6299db9ec823/crc-1.8.1/build.rs --crate-type bin --emit=dep-info,link -C opt-level=3 --cfg 'feature="default"' --cfg 'feature="std"' -C metadata=9eec806c9cb2cdca -C extra-filename=-9eec806c9cb2cdca --out-dir /home/jistone/rust/miniz_oxide/target/release/build/crc-9eec806c9cb2cdca -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --extern build_const=/home/jistone/rust/miniz_oxide/target/release/deps/libbuild_const-237bc424c8fc27b7.rlib --cap-lints allow`
     Running `/home/jistone/rust/miniz_oxide/target/release/build/crc-9eec806c9cb2cdca/build-script-build`
     Running `rustc --crate-name crc /home/jistone/.cargo/registry/src/github.com-1ecc6299db9ec823/crc-1.8.1/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C panic=abort --cfg 'feature="default"' --cfg 'feature="std"' -C metadata=db9c267dd054f996 -C extra-filename=-db9c267dd054f996 --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --cap-lints allow`
   Compiling miniz_oxide v0.1.3 (file:///home/jistone/rust/miniz_oxide/miniz_oxide)
     Running `rustc --crate-name miniz_oxide miniz_oxide/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C panic=abort -C metadata=da63d85bc3dac2b9 -C extra-filename=-da63d85bc3dac2b9 --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --extern adler32=/home/jistone/rust/miniz_oxide/target/release/deps/libadler32-6582e487cd20c9f9.rlib --extern libc=/home/jistone/rust/miniz_oxide/target/release/deps/liblibc-583e369a8ae79b43.rlib`
   Compiling miniz_oxide_c_api v0.1.3 (file:///home/jistone/rust/miniz_oxide)
     Running `rustc --crate-name build_script_build src/build.rs --crate-type bin --emit=dep-info,link -C opt-level=3 --cfg 'feature="default"' -C metadata=8cb98112f97a7701 -C extra-filename=-8cb98112f97a7701 --out-dir /home/jistone/rust/miniz_oxide/target/release/build/miniz_oxide_c_api-8cb98112f97a7701 -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --extern cc=/home/jistone/rust/miniz_oxide/target/release/deps/libcc-8386dfd55958a0e1.rlib`
     Running `/home/jistone/rust/miniz_oxide/target/release/build/miniz_oxide_c_api-8cb98112f97a7701/build-script-build`
     Running `rustc --crate-name miniz_oxide_c_api src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C panic=abort --cfg 'feature="default"' -C metadata=c8ec649d83efdc32 -C extra-filename=-c8ec649d83efdc32 --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --extern crc=/home/jistone/rust/miniz_oxide/target/release/deps/libcrc-db9c267dd054f996.rlib --extern libc=/home/jistone/rust/miniz_oxide/target/release/deps/liblibc-583e369a8ae79b43.rlib --extern miniz_oxide=/home/jistone/rust/miniz_oxide/target/release/deps/libminiz_oxide-da63d85bc3dac2b9.rlib`
    Finished release [optimized] target(s) in 3.32s

$ cargo +nightly test --release --verbose
       Fresh build_const v0.2.1
       Fresh cc v1.0.17
   Compiling libc v0.2.42
   Compiling adler32 v1.0.3
   Compiling crc v1.8.1
     Running `rustc --crate-name adler32 /home/jistone/.cargo/registry/src/github.com-1ecc6299db9ec823/adler32-1.0.3/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C metadata=995d235a15de691d -C extra-filename=-995d235a15de691d --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --cap-lints allow`
   Compiling miniz_oxide_c_api v0.1.3 (file:///home/jistone/rust/miniz_oxide)
     Running `rustc --crate-name libc /home/jistone/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.42/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 --cfg 'feature="default"' --cfg 'feature="use_std"' -C metadata=6e0a133a0f1a257e -C extra-filename=-6e0a133a0f1a257e --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --cap-lints allow`
     Running `rustc --crate-name crc /home/jistone/.cargo/registry/src/github.com-1ecc6299db9ec823/crc-1.8.1/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 --cfg 'feature="default"' --cfg 'feature="std"' -C metadata=0a0ba9a6b5735bbc -C extra-filename=-0a0ba9a6b5735bbc --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --cap-lints allow`
     Running `rustc --crate-name miniz_oxide_c_api src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 --cfg 'feature="default"' -C metadata=829895175cd8b7e5 -C extra-filename=-829895175cd8b7e5 --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --extern crc=/home/jistone/rust/miniz_oxide/target/release/deps/libcrc-db9c267dd054f996.rlib --extern libc=/home/jistone/rust/miniz_oxide/target/release/deps/liblibc-583e369a8ae79b43.rlib --extern miniz_oxide=/home/jistone/rust/miniz_oxide/target/release/deps/libminiz_oxide-da63d85bc3dac2b9.rlib`
   Compiling miniz_oxide v0.1.3 (file:///home/jistone/rust/miniz_oxide/miniz_oxide)
     Running `rustc --crate-name miniz_oxide miniz_oxide/src/lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C metadata=8736d1ee18e2a046 -C extra-filename=-8736d1ee18e2a046 --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --extern adler32=/home/jistone/rust/miniz_oxide/target/release/deps/libadler32-995d235a15de691d.rlib --extern libc=/home/jistone/rust/miniz_oxide/target/release/deps/liblibc-6e0a133a0f1a257e.rlib`
     Running `rustc --crate-name test tests/test.rs --emit=dep-info,link -C opt-level=3 --test --cfg 'feature="default"' -C metadata=69128140fba22fe4 -C extra-filename=-69128140fba22fe4 --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --extern crc=/home/jistone/rust/miniz_oxide/target/release/deps/libcrc-0a0ba9a6b5735bbc.rlib --extern libc=/home/jistone/rust/miniz_oxide/target/release/deps/liblibc-6e0a133a0f1a257e.rlib --extern miniz_oxide=/home/jistone/rust/miniz_oxide/target/release/deps/libminiz_oxide-8736d1ee18e2a046.rlib --extern miniz_oxide_c_api=/home/jistone/rust/miniz_oxide/target/release/deps/libminiz_oxide_c_api-829895175cd8b7e5.rlib`
     Running `rustc --crate-name miniz_oxide_c_api src/lib.rs --emit=dep-info,link -C opt-level=3 --test --cfg 'feature="default"' -C metadata=4e2159d4b5622e18 -C extra-filename=-4e2159d4b5622e18 --out-dir /home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --extern crc=/home/jistone/rust/miniz_oxide/target/release/deps/libcrc-0a0ba9a6b5735bbc.rlib --extern libc=/home/jistone/rust/miniz_oxide/target/release/deps/liblibc-6e0a133a0f1a257e.rlib --extern miniz_oxide=/home/jistone/rust/miniz_oxide/target/release/deps/libminiz_oxide-8736d1ee18e2a046.rlib`
    Finished release [optimized] target(s) in 2.36s
     Running `/home/jistone/rust/miniz_oxide/target/release/deps/miniz_oxide_c_api-4e2159d4b5622e18`

running 3 tests
test tinfl::test::mem_to_heap ... ok
test tinfl::test::mem_to_mem ... ok
test tdef::test::mem_to_heap ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running `/home/jistone/rust/miniz_oxide/target/release/deps/test-69128140fba22fe4`

running 2 tests
test roundtrip_level_1 ... ok
test roundtrip ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests miniz_oxide_c_api
     Running `rustdoc --test /home/jistone/rust/miniz_oxide/src/lib.rs --crate-name miniz_oxide_c_api -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps -L dependency=/home/jistone/rust/miniz_oxide/target/release/deps --cfg 'feature="default"' --extern crc=/home/jistone/rust/miniz_oxide/target/release/deps/libcrc-0a0ba9a6b5735bbc.rlib --extern libc=/home/jistone/rust/miniz_oxide/target/release/deps/liblibc-583e369a8ae79b43.rlib --extern miniz_oxide_c_api=/home/jistone/rust/miniz_oxide/target/release/deps/libminiz_oxide_c_api-829895175cd8b7e5.rlib --extern miniz_oxide=/home/jistone/rust/miniz_oxide/target/release/deps/libminiz_oxide-da63d85bc3dac2b9.rlib --extern libc=/home/jistone/rust/miniz_oxide/target/release/deps/liblibc-6e0a133a0f1a257e.rlib --extern miniz_oxide=/home/jistone/rust/miniz_oxide/target/release/deps/libminiz_oxide-8736d1ee18e2a046.rlib --extern crc=/home/jistone/rust/miniz_oxide/target/release/deps/libcrc-db9c267dd054f996.rlib`
error[E0465]: multiple rlib candidates for `crc` found
 --> /home/jistone/rust/miniz_oxide/src/lib.rs:1:1
  |
1 | extern crate crc;
  | ^^^^^^^^^^^^^^^^^
  |
note: candidate #1: /home/jistone/rust/miniz_oxide/target/release/deps/libcrc-0a0ba9a6b5735bbc.rlib
 --> /home/jistone/rust/miniz_oxide/src/lib.rs:1:1
  |
1 | extern crate crc;
  | ^^^^^^^^^^^^^^^^^
note: candidate #2: /home/jistone/rust/miniz_oxide/target/release/deps/libcrc-db9c267dd054f996.rlib
 --> /home/jistone/rust/miniz_oxide/src/lib.rs:1:1
  |
1 | extern crate crc;
  | ^^^^^^^^^^^^^^^^^

error[E0463]: can't find crate for `crc`
 --> /home/jistone/rust/miniz_oxide/src/lib.rs:1:1
  |
1 | extern crate crc;
  | ^^^^^^^^^^^^^^^^^ can't find crate

error: test failed, to rerun pass '--doc'
```

This does not happen with plain `cargo test`.  We generally use `cargo test --release` in Fedora rpm builds to try and test the same artifacts produced in the rpm `%build` step, but it looks like `panic = "abort"` is throwing a wrench in things.

--- a/src/cargo/core/compiler/compilation.rs
+++ b/src/cargo/core/compiler/compilation.rs
@@ -50,7 +50,9 @@ pub struct Compilation<'cfg> {
     /// be passed to future invocations of programs.
     pub extra_env: HashMap<PackageId, Vec<(String, String)>>,
 
-    pub to_doc_test: Vec<Package>,
+    /// Libraries to test with rustdoc.
+    /// The third value is the list of dependencies.
+    pub to_doc_test: Vec<(Package, Target, Vec<(Target, PathBuf)>)>,
 
     /// Features per package enabled during this compilation.
     pub cfgs: HashMap<PackageId, HashSet<String>>,

--- a/src/cargo/core/compiler/compilation.rs
+++ b/src/cargo/core/compiler/compilation.rs
@@ -9,10 +9,21 @@ use core::{Feature, Package, PackageId, Target, TargetKind};
 use util::{self, join_paths, process, CargoResult, Config, ProcessBuilder};
 use super::BuildContext;
 
+pub struct Doctest {
+    /// The package being doctested.
+    pub package: Package,
+    /// The target being tested (currently always the package's lib).
+    pub target: Target,
+    /// Extern dependencies needed by `rustdoc`. The path is the location of
+    /// the compiled lib.
+    pub deps: Vec<(Target, PathBuf)>,
+}
+
 /// A structure returning the result of a compilation.
 pub struct Compilation<'cfg> {
     /// A mapping from a package to the list of libraries that need to be
     /// linked when working with that package.
+    // TODO: deprecated, remove
     pub libraries: HashMap<PackageId, HashSet<(Target, PathBuf)>>,
 
     /// An array of all tests created during this compilation.
@@ -51,8 +62,7 @@ pub struct Compilation<'cfg> {
     pub extra_env: HashMap<PackageId, Vec<(String, String)>>,
 
     /// Libraries to test with rustdoc.
-    /// The third value is the list of dependencies.
-    pub to_doc_test: Vec<(Package, Target, Vec<(Target, PathBuf)>)>,
+    pub to_doc_test: Vec<Doctest>,
 
     /// Features per package enabled during this compilation.
     pub cfgs: HashMap<PackageId, HashSet<String>>,

--- a/src/cargo/core/compiler/context/mod.rs
+++ b/src/cargo/core/compiler/context/mod.rs
@@ -175,7 +175,7 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
                     None => &output.path,
                 };
 
-                if unit.mode.is_any_test() && !unit.mode.is_check() {
+                if unit.mode == CompileMode::Test {
                     self.compilation.tests.push((
                         unit.pkg.clone(),
                         unit.target.kind().clone(),
@@ -225,6 +225,25 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
                             .iter()
                             .map(|output| (dep.target.clone(), output.path.clone())),
                     );
+            }
+
+            if unit.mode == CompileMode::Doctest {
+                let mut doctest_deps = Vec::new();
+                for dep in self.dep_targets(unit) {
+                    if dep.target.is_lib() && dep.mode == CompileMode::Build {
+                        let outputs = self.outputs(&dep)?;
+                        doctest_deps.extend(
+                            outputs
+                                .iter()
+                                .map(|output| (dep.target.clone(), output.path.clone())),
+                        );
+                    }
+                }
+                self.compilation.to_doc_test.push((
+                    unit.pkg.clone(),
+                    unit.target.clone(),
+                    doctest_deps,
+                ));
             }
 
             let feats = self.bcx.resolve.features(unit.pkg.package_id());

--- a/src/cargo/core/compiler/context/mod.rs
+++ b/src/cargo/core/compiler/context/mod.rs
@@ -1,5 +1,6 @@
 #![allow(deprecated)]
 use std::collections::{HashMap, HashSet};
+use std::ffi::OsStr;
 use std::fmt::Write;
 use std::path::PathBuf;
 use std::sync::Arc;
@@ -8,6 +9,7 @@ use std::cmp::Ordering;
 use jobserver::Client;
 
 use core::{Package, PackageId, Resolve, Target};
+use core::compiler::compilation;
 use core::profiles::Profile;
 use util::errors::{CargoResult, CargoResultExt};
 use util::{internal, profile, Config, short_hash};
@@ -228,6 +230,16 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
             }
 
             if unit.mode == CompileMode::Doctest {
+                // Note that we can *only* doctest rlib outputs here.  A
+                // staticlib output cannot be linked by the compiler (it just
+                // doesn't do that). A dylib output, however, can be linked by
+                // the compiler, but will always fail. Currently all dylibs are
+                // built as "static dylibs" where the standard library is
+                // statically linked into the dylib. The doc tests fail,
+                // however, for now as they try to link the standard library
+                // dynamically as well, causing problems. As a result we only
+                // pass `--extern` for rlib deps and skip out on all other
+                // artifacts.
                 let mut doctest_deps = Vec::new();
                 for dep in self.dep_targets(unit) {
                     if dep.target.is_lib() && dep.mode == CompileMode::Build {
@@ -235,15 +247,19 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
                         doctest_deps.extend(
                             outputs
                                 .iter()
+                                .filter(|output| {
+                                    output.path.extension() == Some(OsStr::new("rlib"))
+                                        || dep.target.for_host()
+                                })
                                 .map(|output| (dep.target.clone(), output.path.clone())),
                         );
                     }
                 }
-                self.compilation.to_doc_test.push((
-                    unit.pkg.clone(),
-                    unit.target.clone(),
-                    doctest_deps,
-                ));
+                self.compilation.to_doc_test.push(compilation::Doctest {
+                    package: unit.pkg.clone(),
+                    target: unit.target.clone(),
+                    deps: doctest_deps,
+                });
             }
 
             let feats = self.bcx.resolve.features(unit.pkg.package_id());

--- a/src/cargo/core/compiler/context/unit_dependencies.rs
+++ b/src/cargo/core/compiler/context/unit_dependencies.rs
@@ -199,19 +199,11 @@ fn compute_deps_custom_build<'a, 'cfg>(
     // 1. Compiling the build script itself
     // 2. For each immediate dependency of our package which has a `links`
     //    key, the execution of that build script.
-    let not_custom_build = unit.pkg
-        .targets()
+    let deps = deps
         .iter()
-        .find(|t| !t.is_custom_build())
-        .unwrap();
-    let tmp = Unit {
-        pkg: unit.pkg,
-        target: not_custom_build,
-        profile: unit.profile,
-        kind: unit.kind,
-        mode: CompileMode::Build,
-    };
-    let deps = deps_of(&tmp, bcx, deps, ProfileFor::Any)?;
+        .find(|(key, _deps)| key.pkg == unit.pkg && !key.target.is_custom_build())
+        .expect("can't find package deps")
+        .1;
     Ok(deps.iter()
         .filter_map(|unit| {
             if !unit.target.linkable() || unit.pkg.manifest().links().is_none() {

--- a/src/cargo/core/compiler/job_queue.rs
+++ b/src/cargo/core/compiler/job_queue.rs
@@ -410,8 +410,6 @@ impl<'a> JobQueue<'a> {
     ) -> CargoResult<()> {
         if (self.compiled.contains(key.pkg) && !key.mode.is_doc())
             || (self.documented.contains(key.pkg) && key.mode.is_doc())
-            // Skip doctest, it is a dummy entry that is always fresh.
-            || key.mode == CompileMode::Doctest
         {
             return Ok(());
         }
@@ -421,8 +419,11 @@ impl<'a> JobQueue<'a> {
             // being a compiled package
             Dirty => {
                 if key.mode.is_doc() {
-                    self.documented.insert(key.pkg);
-                    config.shell().status("Documenting", key.pkg)?;
+                    // Skip Doctest
+                    if !key.mode.is_any_test() {
+                        self.documented.insert(key.pkg);
+                        config.shell().status("Documenting", key.pkg)?;
+                    }
                 } else {
                     self.compiled.insert(key.pkg);
                     if key.mode.is_check() {
@@ -432,11 +433,15 @@ impl<'a> JobQueue<'a> {
                     }
                 }
             }
-            Fresh if self.counts[key.pkg] == 0 => {
-                self.compiled.insert(key.pkg);
-                config.shell().verbose(|c| c.status("Fresh", key.pkg))?;
+            Fresh => {
+                // If doctest is last, only print "Fresh" if nothing has been printed.
+                if self.counts[key.pkg] == 0
+                    && !(key.mode == CompileMode::Doctest && self.compiled.contains(key.pkg))
+                {
+                    self.compiled.insert(key.pkg);
+                    config.shell().verbose(|c| c.status("Fresh", key.pkg))?;
+                }
             }
-            Fresh => {}
         }
         Ok(())
     }

--- a/src/cargo/core/compiler/job_queue.rs
+++ b/src/cargo/core/compiler/job_queue.rs
@@ -410,6 +410,8 @@ impl<'a> JobQueue<'a> {
     ) -> CargoResult<()> {
         if (self.compiled.contains(key.pkg) && !key.mode.is_doc())
             || (self.documented.contains(key.pkg) && key.mode.is_doc())
+            // Skip doctest, it is a dummy entry that is always fresh.
+            || key.mode == CompileMode::Doctest
         {
             return Ok(());
         }
@@ -419,11 +421,8 @@ impl<'a> JobQueue<'a> {
             // being a compiled package
             Dirty => {
                 if key.mode.is_doc() {
-                    // Skip Doctest
-                    if !key.mode.is_any_test() {
-                        self.documented.insert(key.pkg);
-                        config.shell().status("Documenting", key.pkg)?;
-                    }
+                    self.documented.insert(key.pkg);
+                    config.shell().status("Documenting", key.pkg)?;
                 } else {
                     self.compiled.insert(key.pkg);
                     if key.mode.is_check() {

--- a/src/cargo/core/compiler/mod.rs
+++ b/src/cargo/core/compiler/mod.rs
@@ -24,7 +24,7 @@ use self::output_depinfo::output_depinfo;
 
 pub use self::build_context::{BuildContext, FileFlavor, TargetConfig, TargetInfo};
 pub use self::build_config::{BuildConfig, CompileMode, MessageFormat};
-pub use self::compilation::Compilation;
+pub use self::compilation::{Compilation, Doctest};
 pub use self::context::{Context, Unit};
 pub use self::custom_build::{BuildMap, BuildOutput, BuildScripts};
 pub use self::layout::is_bad_artifact_name;

--- a/src/cargo/ops/cargo_compile.rs
+++ b/src/cargo/ops/cargo_compile.rs
@@ -280,7 +280,7 @@ pub fn compile_ws<'a>(
         extra_compiler_args = Some((units[0], args));
     }
 
-    let mut ret = {
+    let ret = {
         let _p = profile::start("compiling");
         let bcx = BuildContext::new(
             ws,
@@ -294,8 +294,6 @@ pub fn compile_ws<'a>(
         let mut cx = Context::new(config, &bcx)?;
         cx.compile(&units, export_dir.clone(), &exec)?
     };
-
-    ret.to_doc_test = to_builds.into_iter().cloned().collect();
 
     return Ok(ret);
 }
@@ -541,9 +539,9 @@ fn generate_targets<'a>(
                     .collect::<Vec<_>>();
                 proposals.extend(default_units);
                 if build_config.mode == CompileMode::Test {
-                    // Include the lib as it will be required for doctests.
+                    // Include doctest for lib.
                     if let Some(t) = pkg.targets().iter().find(|t| t.is_lib() && t.doctested()) {
-                        proposals.push((new_unit(pkg, t, CompileMode::Build), false));
+                        proposals.push((new_unit(pkg, t, CompileMode::Doctest), false));
                     }
                 }
             }
@@ -681,15 +679,7 @@ fn generate_default_targets(targets: &[Target], mode: CompileMode) -> Vec<&Targe
                 })
                 .collect()
         }
-        CompileMode::Doctest => {
-            // `test --doc``
-            targets
-                .iter()
-                .find(|t| t.is_lib() && t.doctested())
-                .into_iter()
-                .collect()
-        }
-        CompileMode::RunCustomBuild => panic!("Invalid mode"),
+        CompileMode::Doctest | CompileMode::RunCustomBuild => panic!("Invalid mode {:?}", mode),
     }
 }
 

--- a/src/cargo/ops/cargo_test.rs
+++ b/src/cargo/ops/cargo_test.rs
@@ -1,7 +1,7 @@
-use std::ffi::{OsStr, OsString};
+use std::ffi::OsString;
 
 use ops;
-use core::compiler::Compilation;
+use core::compiler::{Compilation, Doctest};
 use util::{self, CargoTestError, ProcessError, Test};
 use util::errors::CargoResult;
 use core::Workspace;
@@ -154,7 +154,12 @@ fn run_doc_tests(
         return Ok((Test::Doc, errors));
     }
 
-    for (package, target, deps) in &compilation.to_doc_test {
+    for doctest_info in &compilation.to_doc_test {
+        let Doctest {
+            package,
+            target,
+            deps,
+        } = doctest_info;
         config.shell().status("Doc-tests", target.name())?;
         let mut p = compilation.rustdoc_process(package)?;
         p.arg("--test")
@@ -189,19 +194,6 @@ fn run_doc_tests(
         }
 
         for &(ref target, ref lib) in deps.iter() {
-            // Note that we can *only* doctest rlib outputs here.  A
-            // staticlib output cannot be linked by the compiler (it just
-            // doesn't do that). A dylib output, however, can be linked by
-            // the compiler, but will always fail. Currently all dylibs are
-            // built as "static dylibs" where the standard library is
-            // statically linked into the dylib. The doc tests fail,
-            // however, for now as they try to link the standard library
-            // dynamically as well, causing problems. As a result we only
-            // pass `--extern` for rlib deps and skip out on all other
-            // artifacts.
-            if lib.extension() != Some(OsStr::new("rlib")) && !target.for_host() {
-                continue;
-            }
             let mut arg = OsString::from(target.crate_name());
             arg.push("=");
             arg.push(lib);

--- a/src/cargo/ops/cargo_test.rs
+++ b/src/cargo/ops/cargo_test.rs
@@ -154,86 +154,72 @@ fn run_doc_tests(
         return Ok((Test::Doc, errors));
     }
 
-    let libs = compilation.to_doc_test.iter().map(|package| {
-        (
-            package,
-            package
-                .targets()
-                .iter()
-                .filter(|t| t.doctested())
-                .map(|t| (t.src_path(), t.name(), t.crate_name())),
-        )
-    });
+    for (package, target, deps) in &compilation.to_doc_test {
+        config.shell().status("Doc-tests", target.name())?;
+        let mut p = compilation.rustdoc_process(package)?;
+        p.arg("--test")
+            .arg(target.src_path())
+            .arg("--crate-name")
+            .arg(&target.crate_name());
 
-    for (package, tests) in libs {
-        for (lib, name, crate_name) in tests {
-            config.shell().status("Doc-tests", name)?;
-            let mut p = compilation.rustdoc_process(package)?;
-            p.arg("--test")
-                .arg(lib)
-                .arg("--crate-name")
-                .arg(&crate_name);
+        for &rust_dep in &[&compilation.deps_output] {
+            let mut arg = OsString::from("dependency=");
+            arg.push(rust_dep);
+            p.arg("-L").arg(arg);
+        }
 
-            for &rust_dep in &[&compilation.deps_output] {
-                let mut arg = OsString::from("dependency=");
-                arg.push(rust_dep);
-                p.arg("-L").arg(arg);
+        for native_dep in compilation.native_dirs.iter() {
+            p.arg("-L").arg(native_dep);
+        }
+
+        for &host_rust_dep in &[&compilation.host_deps_output] {
+            let mut arg = OsString::from("dependency=");
+            arg.push(host_rust_dep);
+            p.arg("-L").arg(arg);
+        }
+
+        for arg in test_args {
+            p.arg("--test-args").arg(arg);
+        }
+
+        if let Some(cfgs) = compilation.cfgs.get(package.package_id()) {
+            for cfg in cfgs.iter() {
+                p.arg("--cfg").arg(cfg);
             }
+        }
 
-            for native_dep in compilation.native_dirs.iter() {
-                p.arg("-L").arg(native_dep);
+        for &(ref target, ref lib) in deps.iter() {
+            // Note that we can *only* doctest rlib outputs here.  A
+            // staticlib output cannot be linked by the compiler (it just
+            // doesn't do that). A dylib output, however, can be linked by
+            // the compiler, but will always fail. Currently all dylibs are
+            // built as "static dylibs" where the standard library is
+            // statically linked into the dylib. The doc tests fail,
+            // however, for now as they try to link the standard library
+            // dynamically as well, causing problems. As a result we only
+            // pass `--extern` for rlib deps and skip out on all other
+            // artifacts.
+            if lib.extension() != Some(OsStr::new("rlib")) && !target.for_host() {
+                continue;
             }
+            let mut arg = OsString::from(target.crate_name());
+            arg.push("=");
+            arg.push(lib);
+            p.arg("--extern").arg(&arg);
+        }
 
-            for &host_rust_dep in &[&compilation.host_deps_output] {
-                let mut arg = OsString::from("dependency=");
-                arg.push(host_rust_dep);
-                p.arg("-L").arg(arg);
-            }
+        if let Some(flags) = compilation.rustdocflags.get(package.package_id()) {
+            p.args(flags);
+        }
 
-            for arg in test_args {
-                p.arg("--test-args").arg(arg);
-            }
-
-            if let Some(cfgs) = compilation.cfgs.get(package.package_id()) {
-                for cfg in cfgs.iter() {
-                    p.arg("--cfg").arg(cfg);
-                }
-            }
-
-            let libs = &compilation.libraries[package.package_id()];
-            for &(ref target, ref lib) in libs.iter() {
-                // Note that we can *only* doctest rlib outputs here.  A
-                // staticlib output cannot be linked by the compiler (it just
-                // doesn't do that). A dylib output, however, can be linked by
-                // the compiler, but will always fail. Currently all dylibs are
-                // built as "static dylibs" where the standard library is
-                // statically linked into the dylib. The doc tests fail,
-                // however, for now as they try to link the standard library
-                // dynamically as well, causing problems. As a result we only
-                // pass `--extern` for rlib deps and skip out on all other
-                // artifacts.
-                if lib.extension() != Some(OsStr::new("rlib")) && !target.for_host() {
-                    continue;
-                }
-                let mut arg = OsString::from(target.crate_name());
-                arg.push("=");
-                arg.push(lib);
-                p.arg("--extern").arg(&arg);
-            }
-
-            if let Some(flags) = compilation.rustdocflags.get(package.package_id()) {
-                p.args(flags);
-            }
-
-            config
-                .shell()
-                .verbose(|shell| shell.status("Running", p.to_string()))?;
-            if let Err(e) = p.exec() {
-                let e = e.downcast::<ProcessError>()?;
-                errors.push(e);
-                if !options.no_fail_fast {
-                    return Ok((Test::Doc, errors));
-                }
+        config
+            .shell()
+            .verbose(|shell| shell.status("Running", p.to_string()))?;
+        if let Err(e) = p.exec() {
+            let e = e.downcast::<ProcessError>()?;
+            errors.push(e);
+            if !options.no_fail_fast {
+                return Ok((Test::Doc, errors));
             }
         }
     }

--- a/tests/testsuite/build_script.rs
+++ b/tests/testsuite/build_script.rs
@@ -3044,6 +3044,7 @@ fn panic_abort_with_build_scripts() {
             "src/lib.rs",
             "#[allow(unused_extern_crates)] extern crate a;",
         )
+        .file("build.rs","fn main() {}")
         .file(
             "a/Cargo.toml",
             r#"
@@ -3077,6 +3078,11 @@ fn panic_abort_with_build_scripts() {
     assert_that(
         p.cargo("build").arg("-v").arg("--release"),
         execs().with_status(0),
+    );
+
+    assert_that(
+        p.cargo("test --release"),
+        execs().with_status(0)
     );
 }
 

